Chunk taskArns into groups of 100

owen opened the pull request:
If the ECS cluster has > 100 tasks, passing them to
ecs.DescribeTasksRequest() will result in the AWS API returning
errors.

This patch breaks them into chunks of at most 100, and calls
DescribeTasks for each chunk.

timoreimann commented on provider/ecs.go:
```diff
@@ -206,13 +206,30 @@ func (provider *ECS) listInstances(ctx context.Context, client *awsClient) ([]ec
+			sliceEnd = len(taskArns)
+		}
+		split_taskreq = append(splitTaskReq, taskArns[i:sliceEnd])
+	}
```
Can we move the splitting logic into a small function, e.g.,

```go
func chunkedTaskArns(taskArns []*string) [][]*string
```

?

timoreimann replied:
Some tests on the newly created function would also be nice to cover the representative cases (e.g., 0, 1, 99, 100, 101, 200, 400).

owen replied:
good thinking!  I'll catch that :)

owen replied:
also, my git-fu is weak and I managed to commit a broken patch :(  fixes soon.

emilevauge commented:
@owen shouldn't we rebase this one `v1.2` branch to include it in the next release?

timoreimann commented:
@owen how's your git-fu going today? ;)

also, do you want to rebase against 1.2 as @emilevauge suggested? The bugfix seems fairly important for ECS users.

owen commented:
it's _woeful!_ I'm king of the force push I guess.

I've: 
 * rebased against v1.2
 * split to a small helper function
 * added test cases around splittable values

and will be force pushing again as soon as I ensure it passes a `make validate`

Should I keep a PR against master with this code, or do you all typically pull v1.2 fixes back in?  not sure how that typically works for this project.
(and oh look, more git questions, hehe)

timoreimann commented:
@owen we typically sync back the v1.2 branch with master regularly (usually on release of a release candidate / final version), so there should be no need to maintain a separate branch against master.

piteur commented on provider/ecs.go:
```diff
@@ -206,13 +206,30 @@ func (provider *ECS) listInstances(ctx context.Context, client *awsClient) ([]ec
-		Tasks:   taskArns,
-		Cluster: &provider.Cluster,
-	})
+	var splitTaskReq [][]*string
```
Simple comment - as an outsider interrested on this fix:

Is adding a comment to explain why we split the content & do multiple calls will add any value for future maintenance ?
Without knowing the API limitation - or not having the PR context in mind - the `100` split value can be interpreted as coming from nowhere / seems a bit magical.

timoreimann replied:
+1 on that one; plus, moving the logic into a dedicated function with an informative name ([as suggested previously](https://github.com/containous/traefik/pull/1209#discussion_r103832525) :-) )

owen commented:
A test case failure I observed last night made me realize that pointing Traefik to an empty cluster would result in bad AWS API calls being made.  I've gone ahead and picked up that fix by adding an early return after the first call to ListTasks.

This now works for my empty cluster, and my ~hundred instance cluster.  Tests pass, `make validate` looks good, and it only took _one more force push_ to get it to go up 😬 

@timoreimann

owen commented:
@timoreimann / @emilevauge - I'm wondering if I broke your CI environment with my, er, aggressive Git usage.

Let me know if I can help :(

emilevauge commented:
@owen yeah, it seems Travis needs a holiday... Could you push force again?

owen commented:
Looks like the CI is building now.  I added a few more tests to wake up the CI pipeline.

Apologies... I actually just learned that "rebase master -> v1.2" is as easy as a click at the top of the pull request.  That would've skipped at least one force push on my end, heh.
(I have much to learn about GitHub ;-))

timoreimann commented on provider/ecs_test.go:
```diff
@@ -308,3 +308,42 @@ func TestFilterInstance(t *testing.T) {
+		}
+
+		if !reflect.DeepEqual(outCount, c.expectedLengths) {
+			t.Fatalf("Chunking %d elements, expected %#v, got %#v (case %d)", c.count, c.expectedLengths, outCount, i)
```
I think we should do `t.Errorf` here: If one case fails, we still want to see the result of the other cases.

timoreimann replied:
I also think we can drop the index variable `i` here. `c.count` already suffices to identify the failing case unambiguously.

owen replied:
Not sure why I `Fatalf`'ed there rather than `Errorf`'ed - fixed :)

owen commented on provider/ecs.go:
```diff
@@ -206,13 +206,24 @@ func (provider *ECS) listInstances(ctx context.Context, client *awsClient) ([]ec
-		Tasks:   taskArns,
-		Cluster: &provider.Cluster,
-	})
+	if len(taskArns) == 0 {
```
👍

timoreimann commented:
@owen you haven't go fmt'ed your latest commit -- please run `make fmt` to fix that. Apart from that, LGTM! :heart:

Could you also squash your commits? After that and another successful CI run, I think we are good to go/merge. :-)

owen commented:
[I fear Travis may not be happy with me again](https://travis-ci.org/containous/traefik/jobs/208818157):

```Unable to checkout gopkg.in/fsnotify.v1
[ERROR]	Update failed for gopkg.in/fsnotify.v1: Cannot detect VCS
[ERROR]	Failed to install: Cannot detect VCS
The command '/bin/sh -c cd integration && glide install' returned a non-zero code: 1
make: *** [build] Error 1
```

timoreimann commented:
Github is flaking. Retriggered the build.

timoreimann commented:
Grrr, and now the integration tests are flaking. :(

Retrying HARDER...

emilevauge commented:
![](https://i.imgflip.com/1l1f6h.jpg)
